src/memory/array.rs

```diff
@@ -0,0 +1,545 @@
+    }
+
+    /// Returns a copy of `self` with the LAYERED flag set
+    pub fn mark_layered(self) -> ArrayObjectFlags {
+        self | ArrayObjectFlags::LAYERED
+    }
+
+    /// Checks if SURFACE_LDST flag is set
+    pub fn surface_ldst(self) -> bool {
+        self.contains(ArrayObjectFlags::SURFACE_LDST)
+    }
+
+    /// Returns a copy of `self` with the SURFACE_LDST flag set
+    pub fn mark_surface_ldst(self) -> ArrayObjectFlags {
+        self | ArrayObjectFlags::SURFACE_LDST
+    }
+
+    /// Checks if CUBEMAP flag is set
+    pub fn cubemap(self) -> bool {
+        self.contains(ArrayObjectFlags::CUBEMAP)
+    }
+
+    /// Returns a copy of `self` with the CUBEMAP flag set
+    pub fn mark_cubemap(self) -> ArrayObjectFlags {
+        self | ArrayObjectFlags::CUBEMAP
+    }
+
+    /// Checks if TEXTURE_GATHER flag is set
+    pub fn texture_gather(self) -> bool {
+        self.contains(ArrayObjectFlags::TEXTURE_GATHER)
+    }
+
+    /// Returns a copy of `self` with the TEXTURE_GATHER flag set
+    pub fn mark_texture_gather(self) -> ArrayObjectFlags {
+        self | ArrayObjectFlags::TEXTURE_GATHER
+    }
+}
+
+/// Describes a CUDA Array
+#[derive(Clone, Copy, Debug)]
+pub struct ArrayDescriptor {
+    desc: cuda_sys::cuda::CUDA_ARRAY3D_DESCRIPTOR,
+}
+
+impl ArrayDescriptor {
+    /// Constructs an ArrayDescriptor from a CUDA Driver API Array Descriptor.
+    pub fn from_raw(desc: cuda_sys::cuda::CUDA_ARRAY3D_DESCRIPTOR) -> Self {
+        Self { desc }
+    }
+
+    /// Creates a new ArrayDescriptor from a set of dimensions and format.
+    pub fn from_dims_format(dims: [usize; 3], format: ArrayFormat) -> Self {
+        Self {
+            desc: cuda_sys::cuda::CUDA_ARRAY3D_DESCRIPTOR {
+                Width: dims[0],
+                Height: dims[1],
+                Depth: dims[2],
+                Format: format.to_raw(),
+                NumChannels: 1,
+                Flags: ArrayObjectFlags::default().bits(),
+            },
+        }
+    }
+
+    /// Returns the dimensions of the ArrayDescriptor
+    pub fn dims(&self) -> [usize; 3] {
+        [self.desc.Width, self.desc.Height, self.desc.Depth]
+    }
+
+    /// Sets the dimensions of the ArrayDescriptor
+    pub fn set_dims(&mut self, dims: [usize; 3]) {
+        self.desc.Width = dims[0];
+        self.desc.Height = dims[1];
+        self.desc.Depth = dims[2];
+    }
+
+    /// Sets the dimensions of the ArrayDescriptor. Composes with other with_ routines.
+    pub fn with_dims(mut self, dims: [usize; 3]) -> Self {
+        self.set_dims(dims);
+        self
+    }
+
+    /// Returns the width of the ArrayDescripor
+    pub fn width(&self) -> usize {
+        self.desc.Width
+    }
+
+    /// Sets the width of the ArrayDescriptor
+    pub fn set_width(&mut self, width: usize) {
+        self.desc.Width = width;
+    }
+
+    /// Sets the width of the ArrayDescriptor. Composes with other with_ routines.
+    pub fn mark_width(mut self, width: usize) -> Self {
```

Comment:
`mark_width` is a slightly confusing name. I'm not sure that we need builder functions, especially since none of the other types in RustaCUDA have them. I could be convinced either way though. Should be consistent either way.

Reply:
Will remove.